PR: https://github.com/odoo/odoo/pull/259842

From: d069ce59e28fda2bc25fb89fd01ee7b988a30fa0
From: Nisarg (nipl)
Date: 2026-06-15 17:24:30
+    tax_ids = fields.Many2many(string="Taxes", comodel_name="account.tax", check_company=True)

Total Changes: 158

[IMP] sale, account, *: support description-first flow on sale orders and invoices

This commit introduces support for productless Sale Order Lines (SOLs),
allowing users to create quotation/invoice lines directly from descriptions
without requiring a product.

Functional changes

* For small business workflows, creating products for every line can be
  unnecessary overhead. SOLs can now be created without a product while
  preserving standard sale and invoicing flows.
* A new setting allows companies to keep the previous product-first behavior
  by making products mandatory on SOLs.
* When products are mandatory, UoM selection remains restricted to the
  product's compatible UoMs.
* For productless SOLs, users can select any UoM.
* All standard SOLs (`display_type=False`) now default to `Units` as UoM.
  Once a product is selected, the UoM is automatically adjusted according
  to the product configuration.
* New productless SOLs now receive the company default sales taxes by
  default, while still allowing manual tax customization.
* Automatic Invoicing setting is now converted to company specific setting
  instead of config param.
* When applying quotation templates on sale orders now created SOLs retain
  product name + (template) description, for consistancy on both the places
* Productless SOLs are now fully supported in section templates:
  taxes, unit price, discount, and unit cost are preserved when saving
  and reapplying templates (currency conversions are also handled).
* Loyalty programs having no product restriction takes productless SOLs
  in cosideration when calculating rewards.
* In sales reports productless SOLs are displayed as their name instead
  of product name (with selected UOM).
* For Commissions and achievements rules which have no product restrictions
  directly aggregates productless SOLs.
* To keep consistant portal flow, eCommerce links to products are now only
  shown for eCommerce orders. (because productless lines won't have the link
  and it shouldn't be user's concern whether some product is there in our DB
  or not).
* When creating the same product as one already added through a quotation
  template, we previously reused the template's description. This is not
  very useful, as if the user wants to sell the product under the same
  description, they would typically just increase the quantity instead.

UI & UX

* Description field of SO and AM  now acts as the main product + description
  textarea and supports direct product search from the description itself,
  without requiring interaction with the product field.
* When a product is readonly, editing the description no longer opens the
  product autocomplete, ensuring users can safely edit labels without
  modifying the product.
* Sale configurators and combo flows continue to behave as before on new field.
* Any configuration or combo flow using the description field column now
  automatically uses the new searchable description field widget.
* Edit Configuration button is now only visible on hover, for better UX.
* Also stacked description is now always visible, the button for toggle its
  visibility is completely removed.
* When mandatory product feature is active, users won't be able to search on
  description field, for mandatory product mode, users must switch to old
  Product field.
* Edit button UX is changed bit to only show on hover (and when line is not
  edit mode).
* Removed `X {quantity}` from combo product line's display_name since we
  already display quantity field on combo line and it's pointless code to
  maintain. (eg. Office Combo X 3 -> Office Combo)
* Description now completely reset through onchange when user removes
  product from exisiting line.
* For managing productless lines in quotation/section templates we introduced
  optional hidden columns unit price, discount and taxes on template lines,
  which are invisible for lines having product, and should be selectable when
  atleast one productless line is added in template.

Technical changes

* `default_invoice_policy` was converted to `sale_invoice_policy`, a
  company-dependent field.
* For SOL invoice policy:
  * product SOLs use `product_id.invoice_policy`
  * productless SOLs fallback to `company.sale_invoice_policy`
* For searching invoice policy of any SOL (with or without product)
  a new computed search field `sale.order.line.invoice_policy` is introduced.
* Extracted combos and configurators logic in `saleProductMixin` which contains
  methods which should be called after `product.template` or `product.product`
  updates rest of the things are handled in mixin itself.
* Cleaned Up and improved some `sale` tour utils such as
  * reduced indeterministic behaviours by adding a `o_has_product` flag on
    description field.
  * Added logic to check multiline description of some SOL.
  * Refined some selectors of tours to gain better certainity on element
    we want to select.
* We also noticed that `product_id` was not removed in onchange when removing
  product from `product_template_id` field, which caused some issues like
  old name still showing there or having `product_id` w/o `product_template_id`.
  Added onchange to make both the fields consistant.

task-6109825

Part-of: odoo/odoo#259842
Related: odoo/enterprise#114608
Related: odoo/upgrade#9981
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>
Co-authored-by: Nisarg Patel (nipl) <nipl@odoo.com>
Co-authored-by: Fabien Pinckaers (fp) <fp@openerp.com>
Co-authored-by: John Laterre (jol) <jol@odoo.com>
Co-authored-by: Julien Carion (juca) <juca@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale_management/models/sale_order.py
+++ b/addons/sale_management/models/sale_order.py
@@ -74,11 +74,15 @@ class SaleOrder(models.Model):
         if not self.sale_order_template_id:
             return
 
-        sale_order_template = self.sale_order_template_id.with_context(lang=self.partner_id.lang)
+        sale_order_template = self.sale_order_template_id.with_context(
+            lang=self.partner_id.lang
+        ).with_company(self.company_id)
 
         order_lines_data = [fields.Command.clear()]
         order_lines_data += [
-            fields.Command.create(line._prepare_order_line_values())
+            fields.Command.create(
+                line._prepare_order_line_values(self.fiscal_position_id, self.currency_id)
+            )
             for line in sale_order_template.sale_order_template_line_ids
         ]
 

--- a/addons/sale_management/models/sale_order_line.py
+++ b/addons/sale_management/models/sale_order_line.py
@@ -1,6 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models
+from odoo import fields, models
 from odoo.fields import Command, Domain
 
 
@@ -12,35 +12,6 @@ class SaleOrderLine(models.Model):
         string="Optional Line", copy=True, default=False
     )  # Whether this section's lines are optional in the portal.
 
-    # === COMPUTE METHODS === #
-
-    @api.depends("product_id")
-    def _compute_name(self):
-        # Take the description on the order template if the product is present in it
-        super()._compute_name()
-        for line in self:
-            order = line.order_id
-            if line.product_id and order.sale_order_template_id and line._use_template_name():
-                for template_line in order.sale_order_template_id.sale_order_template_line_ids:
-                    if line.product_id == template_line.product_id and template_line.name:
-                        # If a specific description was set on the template, use it
-                        # Otherwise the description is handled by the super call
-                        lang = order.partner_id.lang
-                        line.name = (
-                            template_line.with_context(lang=lang).name
-                            + line.with_context(
-                                lang=lang
-                            )._get_sale_order_line_multiline_description_variants()
-                        )
-                        break
-
-    def _use_template_name(self):
-        """Decide whether the quotation template description should be used for this line (if any
-        template line matches the current one).
-        """
-        self.ensure_one()
-        return True
-
     # === PUBLIC === #
 
     def save_section_template(self):
@@ -78,12 +49,15 @@ class SaleOrderLine(models.Model):
         ]
 
         if existing_template:
-            template_lines_data = [Command.clear(), *template_lines]
+            vals = {"sale_order_template_line_ids": [Command.clear(), *template_lines]}
+            if existing_template.currency_id != self.order_id.currency_id:
+                vals["currency_id"] = self.order_id.currency_id.id
+
             # .sudo because we allow salesman to update their own templates
-            existing_template.sudo().sale_order_template_line_ids = template_lines_data
+            existing_template.sudo().write(vals)
             return existing_template.read(["id", "name", "create_uid"], load="")[0]
 
-        # .sudo because we allow salesman to and maintaincreate their own templates
+        # .sudo because we allow salesman to maintain and create their own templates
         new_template = (
             self
             .env["sale.order.template"]
@@ -93,6 +67,7 @@ class SaleOrderLine(models.Model):
                 "template_type": "section",
                 "sale_order_template_line_ids": template_lines,
                 "company_id": self.order_id.company_id.id,
+                "currency_id": self.order_id.currency_id.id,
                 "share_template": False,
             })
         )
@@ -118,11 +93,14 @@ class SaleOrderLine(models.Model):
     def _prepare_template_line_values(self):
         """Prepare create values for a sale order template line from a sale order line.
 
+        If the line is linked to a product, the product is stored and pricing is recomputed later.
+        For product lines without a product, price, discount, and taxes are copied explicitly.
+
         :return: `sale.order.template.line` create values
         :rtype: dict
         """
         self.ensure_one()
-        return {
+        vals = {
             "name": self.name,
             "product_uom_qty": self.product_uom_qty,
             "product_uom_id": self.product_uom_id.id,
@@ -132,3 +110,15 @@ class SaleOrderLine(models.Model):
             "collapse_composition": self.collapse_composition,
             "collapse_prices": self.collapse_prices,
         }
+
+        if not self.product_id:
+            vals.update({
+                "tax_ids": [Command.set(self.tax_ids.ids)],
+                "discount": self.discount,
+                "price_unit": self.price_unit,
+            })
+        else:
+            # Try to remove the default product display_name from the line name
+            vals["name"] = self.name.removeprefix(f"{self.product_id.display_name}\n")
+
+        return vals

--- a/addons/sale_management/models/sale_order_template.py
+++ b/addons/sale_management/models/sale_order_template.py
@@ -15,6 +15,7 @@ class SaleOrderTemplate(models.Model):
         help="If unchecked, it will allow you to hide the Order template without removing it.",
     )
     company_id = fields.Many2one(comodel_name="res.company", default=lambda self: self.env.company)
+    currency_id = fields.Many2one(string="Currency", comodel_name="res.currency")
 
     name = fields.Char(string="Template", required=True)
     note = fields.Html(string="Terms and conditions", translate=True)
@@ -83,6 +84,7 @@ class SaleOrderTemplate(models.Model):
         compute="_compute_user_has_access",
         search="_search_user_has_access",
     )
+    has_productless_lines = fields.Boolean(compute="_compute_has_productless_lines")
 
     # === COMPUTE METHODS ===#
 
@@ -136,6 +138,14 @@ class SaleOrderTemplate(models.Model):
             )
         ) | Domain("create_uid", x2many_operator, self.env.user.ids)
 
+    @api.depends("sale_order_template_line_ids")
+    def _compute_has_productless_lines(self):
+        for template in self:
+            template.has_productless_lines = any(
+                not (line.product_id or line.display_type)
+                for line in self.sale_order_template_line_ids
+            )
+
     # === ONCHANGE METHODS ===#
 
     @api.onchange("prepayment_percent")
@@ -310,7 +320,9 @@ class SaleOrderTemplate(models.Model):
         )
         return self.search_read(domain, fields=["id", "name", "create_uid"], load="")
 
-    def prepare_section_template_order_lines(self, order_changes, fields_spec):
+    def prepare_section_template_order_lines(
+        self, order_changes, fiscal_position_id, company_id, currency_id, fields_spec
+    ):
         """Prepare `sale.order.line` value dicts from a section template.
 
         Builds order line values from the given section template, applies
@@ -318,15 +330,23 @@ class SaleOrderTemplate(models.Model):
         returns the resulting values ready for insertion.
 
         :param dict order_changes: Order values to consider for onchange
+        :param int fiscal_position_id: fiscal position of the order
+        :param int company_id: company of the order
+        :param int currency_id: currency of the order
         :param dict fields_spec: Fields specification for onchange
         :return: Prepared sale order line values
         :rtype: list[dict]
         """
         self.ensure_one()
         result = []
-
-        for line in self.sale_order_template_line_ids:
-            onchange_values = {**line._prepare_order_line_values(), **order_changes}
+        fiscal_position = self.env["account.fiscal.position"].browse(fiscal_position_id)
+        currency = self.env["res.currency"].browse(currency_id)
+
+        for line in self.with_company(company_id).sale_order_template_line_ids:
+            onchange_values = {
+                **line._prepare_order_line_values(fiscal_position, currency),
+                **order_changes,
+            }
             onchange_result = self.env["sale.order.line"].onchange(onchange_values, [], fields_spec)
             result.append(onchange_result.get("value", {}))
 

--- a/addons/sale_management/models/sale_order_template_line.py
+++ b/addons/sale_management/models/sale_order_template_line.py
@@ -2,6 +2,7 @@
 
 from odoo import api, fields, models
 from odoo.exceptions import UserError
+from odoo.fields import Command
 
 
 class SaleOrderTemplateLine(models.Model):
@@ -10,8 +11,8 @@ class SaleOrderTemplateLine(models.Model):
     _order = "sale_order_template_id, sequence, id"
 
     _accountable_product_id_required = models.Constraint(
-        "CHECK(display_type IS NOT NULL OR (product_id IS NOT NULL AND product_uom_id IS NOT NULL))",  # noqa: E501
-        "Missing required product and UoM on accountable sale quote line.",
+        "CHECK(display_type IS NOT NULL OR product_uom_id IS NOT NULL)",
+        "Missing required UoM on accountable sale quote line.",
     )
     _non_accountable_fields_null = models.Constraint(
         "CHECK(display_type IS NULL OR (product_id IS NULL AND product_uom_qty = 0 AND product_uom_id IS NULL))",  # noqa: E501
@@ -47,7 +48,7 @@ class SaleOrderTemplateLine(models.Model):
     product_uom_id = fields.Many2one(
         comodel_name="uom.uom",
         string="Unit",
-        domain="[('id', 'in', allowed_uom_ids)]",
+        domain="[('id', 'in', allowed_uom_ids)] if allowed_uom_ids or mandatory_product else []",
         compute="_compute_product_uom_id",
         store=True,
         readonly=False,
@@ -72,19 +73,34 @@ class SaleOrderTemplateLine(models.Model):
     collapse_composition = fields.Boolean()
     collapse_prices = fields.Boolean()
 
-    # === COMPUTE METHODS ===#
+    # Technical fields which stores values for product SO line without product_id
+    discount = fields.Float(string="Discount (%)", digits="Discount")
+    price_unit = fields.Float(
+        string="Unit Price", digits="Product Price", min_display_digits="Product Price"
+    )
+    tax_ids = fields.Many2many(string="Taxes", comodel_name="account.tax", check_company=True)
 
-    @api.depends(
-        "product_id", "product_id.uom_id", "product_id.uom_ids", "product_id.extra_uom_ids"
+    mandatory_product = fields.Boolean(
+        string="Is Product Mandatory", compute="_compute_mandatory_product"
     )
+
+    # === COMPUTE METHODS ===#
+
+    @api.depends("product_id")
     def _compute_allowed_uom_ids(self):
         for option in self:
             option.allowed_uom_ids = option.product_id._get_available_uoms()
 
-    @api.depends("product_id")
+    @api.depends("product_id", "display_type")
     def _compute_product_uom_id(self):
+        unit_uom = self.env["product.template"]._default_uom_id()
         for option in self:
-            option.product_uom_id = option.product_id.uom_id
+            if option.display_type:
+                option.product_uom_id = False
+            elif not option.product_id:
+                option.product_uom_id = unit_uom
+            else:
+                option.product_uom_id = option.product_id.uom_id
 
     def _compute_parent_id(self):
         option_lines = set(self)
@@ -107,6 +123,11 @@ class SaleOrderTemplateLine(models.Model):
                 elif line in option_lines:
                     line.parent_id = last_sub or last_section
 
+    def _compute_mandatory_product(self):
+        self.mandatory_product = (
+            self.env["ir.config_parameter"].sudo().get_bool("sale.mandatory_product")
+        )
+
     # === CRUD METHODS ===#
 
     @api.model_create_multi
@@ -135,8 +156,14 @@ class SaleOrderTemplateLine(models.Model):
         """Return the domain of the products that can be added to the template."""
         return [("sale_ok", "=", True), ("type", "!=", "combo")]
 
-    def _prepare_order_line_values(self):
-        """Give the values to create the corresponding order line.
+    def _prepare_order_line_values(self, fiscal_position, currency):
+        """Prepare values to create a sale order line from a template line.
+
+        Line without products take price, discount, taxes from itself otherwise compute it based on
+        product and related values.
+
+        :param account.fiscal.position fiscal_position_id: fiscal position to use
+        :param res.currency currency: target currency (of the order)
 
         :return: `sale.order.line` create values
         :rtype: dict
@@ -154,4 +181,21 @@ class SaleOrderTemplateLine(models.Model):
         }
         if self.name:
             vals["name"] = self.name
+            if self.product_id:
+                vals["name"] = f"{self.product_id.display_name}\n{self.name}"
+
+        if not self.product_id:
+            taxes = self.tax_ids._filter_taxes_by_company()
+
+            if fiscal_position:
+                taxes = fiscal_position.map_tax(taxes)
+
+            vals.update({
+                "tax_ids": [Command.set(taxes.ids)],
+                "discount": self.discount,
+                "price_unit": self.sale_order_template_id.currency_id._convert(
+                    from_amount=self.price_unit, to_currency=currency
+                ),
+            })
+
         return vals
